packages/flutter/lib/src/widgets/restoration_properties.dart

```diff
@@ -0,0 +1,250 @@
+/// [Listenable].
+///
+/// This class may be used to implement a [RestorableProperty] for a
+/// [Listenable], whose information it needs to store in the restoration data
+/// change whenever the [Listenable] notifies its listeners.
+///
+/// The [RestorationMixin] this property is registered with will call
+/// [toPrimitives] whenever the wrapped [Listenable] notifies its listeners to
+/// update the information that this property has stored in the restoration
+/// data.
+abstract class RestorableListenable<T extends Listenable> extends RestorableProperty<T> {
+  /// The [Listenable] stored in this property.
+  ///
+  /// A representation of the current value of the [Listenable] is stored in the
+  /// restoration data. During state restoration, the [Listenable] returned by
+  /// this getter will be restored to the state it had when the restoration data
+  /// the property is getting restored from was collected.
+  ///
+  /// The [value] can only be accessed after the property has been registered
+  /// with a [RestorationMixin] by calling
+  /// [RestorationMixin.registerForRestoration].
+  T get value {
+    assert(isRegistered);
+    return _value;
+  }
+  T _value;
+
+  @override
+  void initWithValue(T value) {
+    assert(value != null);
+    _value?.removeListener(notifyListeners);
+    _value = value;
+    _value.addListener(notifyListeners);
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+    _value?.removeListener(notifyListeners);
+  }
+}
+
+/// A [RestorableProperty] that knows how to store and restore a
+/// [TextEditingController].
+///
+/// The [TextEditingController] is accessible via the [value] getter. During
+/// state restoration, the property will restore [TextEditingController.text] to
+/// the value it had when the restoration data it is getting restored from was
+/// collected.
+class RestorableTextEditingController extends RestorableListenable<TextEditingController> {
+  /// Creates a [RestorableTextEditingController].
+  ///
+  /// This constructor treats a null `text` argument as if it were the empty
+  /// string.
+  factory RestorableTextEditingController({String text}) => RestorableTextEditingController.fromValue(
+    text == null ? TextEditingValue.empty : TextEditingValue(text: text),
+  );
+
+  /// Creates a [RestorableTextEditingController] from an initial
+  /// [TextEditingValue].
+  ///
+  /// This constructor treats a null `value` argument as if it were
+  /// [TextEditingValue.empty].
+  RestorableTextEditingController.fromValue(TextEditingValue value) : _initialValue = value;
+
+  final TextEditingValue _initialValue;
+
+  @override
+  TextEditingController createDefaultValue() {
+    return TextEditingController.fromValue(_initialValue);
+  }
+
+  @override
+  TextEditingController fromPrimitives(Object data) {
+    return TextEditingController(text: data as String);
+  }
+
+  @override
+  Object toPrimitives() {
+    return value.text;
```

Comment:
shouldn't we also save the selection?

Reply:
Saving the selection seems to only make sense if we restore focus, which we are currently not doing.